iterator/cdc.go

```diff
@@ -29,224 +31,254 @@ import (
-	}
-
-	// now that we have the shard ID, we can fetch the shard iterator
-	return c.getShardIteratorForShard(ctx, selectedShardID, shardIteratorType)
+	return nil
 }
 
-// getShardIteratorForShard gets the shard iterator for a specific shard.
-func (c *CDCIterator) getShardIteratorForShard(ctx context.Context, shardID string, shardIteratorType stypes.ShardIteratorType) (*string, error) {
+func (c *CDCIterator) startShardProcessor(ctx context.Context, shard stypes.Shard, sequenceNumber string) (*shardProcessor, error) {
+	shardID := *shard.ShardId
+	var shardIteratorType stypes.ShardIteratorType
+	var startingSequenceNumber *string
+
+	if sequenceNumber != "" {
+		// this is the shard containing the sequence number
+		shardIteratorType = stypes.ShardIteratorTypeAfterSequenceNumber
+		startingSequenceNumber = aws.String(sequenceNumber)
+	} else {
+		// for other shards, use TrimHorizon to read from the beginning
+		shardIteratorType = stypes.ShardIteratorTypeTrimHorizon
+	}
+
 	input := &dynamodbstreams.GetShardIteratorInput{
 		StreamArn:         aws.String(c.streamArn),
 		ShardId:           aws.String(shardID),
 		ShardIteratorType: shardIteratorType,
-	}
-	// continue from a specific position.
-	if shardIteratorType == stypes.ShardIteratorTypeAfterSequenceNumber {
-		input.SequenceNumber = aws.String(c.p.SequenceNumber)
+		SequenceNumber:    startingSequenceNumber,
 	}
 
-	// get the shard iterator.
 	getShardIteratorOutput, err := c.streamsClient.GetShardIterator(ctx, input)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get shard iterator for shard %s: %w", shardID, err)
 	}
 
-	return getShardIteratorOutput.ShardIterator, nil
-}
+	shardProcessor := &shardProcessor{
+		shardID:       shardID,
+		shardIterator: getShardIteratorOutput.ShardIterator,
+		tomb:          &tomb.Tomb{},
+	}
 
-// moveToNextShard used to get the iterator of the shard that follows the current one after it was closed.
-func (c *CDCIterator) moveToNextShard(ctx context.Context) (*string, error) {
-	// describe the stream to get the shard details.
-	describeStreamOutput, err := c.streamsClient.DescribeStream(ctx, &dynamodbstreams.DescribeStreamInput{
-		StreamArn: aws.String(c.streamArn),
+	// start processing this shard
+	shardProcessor.tomb.Go(func() error {
+		return c.processShard(ctx, shardProcessor)
 	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to describe stream: %w", err)
-	}
 
-	shards := describeStreamOutput.StreamDescription.Shards
+	return shardProcessor, nil
+}
 
-	// move to the next shard if the current one is closed
-	if c.shardIndex+1 < len(shards) {
-		c.shardIndex++
-		nextShard := shards[c.shardIndex]
-		// get shard iterator for the new shard
-		return c.getShardIteratorForShard(ctx, *nextShard.ShardId, stypes.ShardIteratorTypeTrimHorizon)
+func (c *CDCIterator) processShard(ctx context.Context, s *shardProcessor) error {
+	defer func() {
+		// clenaup the shard from the SequenceNumberMap when shard is closed.
+		c.cacheLock.Lock()
```

Comment:
What's the reason that we use the cache lock here, and not something specific for the last position?

Reply:
my bad